cloudformation/slack-router-cfn-template.js

```diff
@@ -0,0 +1,94 @@
+const cf = require('@mapbox/cloudfriend')
+
+const Resources = {
+  SlackRouterApi: {
+    Type: 'AWS::ApiGatewayV2::Api',
+    Properties: {
+      Name: 'slack-router-api',
+      ProtocolType: 'HTTP',
+    },
+  },
+  SlackToLambdaIntegration: {
+    Type: 'AWS::ApiGatewayV2::Integration',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+      Description: 'Integration for Slack Slash Commands',
+      IntegrationType: 'AWS_PROXY',
+      IntegrationUri: cf.join('', [
+        'arn:',
+        cf.partition,
+        ':apigateway:',
+        cf.region,
+        ':lambda:path/2015-03-31/functions/',
+        cf.getAtt('SlackRouterLambda', 'Arn'),
+        '/invocations',
+      ]),
+      PayloadFormatVersion: '2.0',
+    },
+  },
+  SlackToLambdaRoute: {
+    Type: 'AWS::ApiGatewayV2::Route',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+      AuthorizationType: 'NONE',
+      RouteKey: 'POST /slack-to-lambda',
+      Target: cf.join('/', [
+        'integrations',
+        cf.ref('SlackToLambdaIntegration'),
+      ]),
+    },
+  },
+  SlackToLambdaDeploy: {
+    Type: 'AWS::ApiGatewayV2::Deployment',
+    DependsOn: 'SlackToLambdaRoute',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+    },
+  },
+  SlackRouterApiProdStage: {
+    Type: 'AWS::ApiGatewayV2::Stage',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+      AutoDeploy: 'true',
+      DeploymentId: cf.ref('SlackToLambdaDeploy'),
+      StageName: 'prod',
+    },
+  },
+  SlackRouterLambda: {
+    Type: 'AWS::Lambda::Function',
+    Properties: {
+      FunctionName: 'slack-router-lambda',
```

Comment:
This will need to be unique- add `cf.stackName` with a cf.join.

Reply:
@dakotabenjamin I'm confused, do I need to add a tag that's unique? Or the function name has to be unique? We wanted the Lambda functions to be the exact same name as its respective SNS which has the same name as the Slash command (so the `/tm-health` command will have a `tm-health` SNS and Lambda).

I would appreciate some guidance on how to do the tags. I see in the [docs here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-resource-tags.html) that CloudFormation automatically creates `aws:cloudformation:stack-name` (among others) as a stack-level tag (which I can see in the Lambda functions although not in the SNS topics) so do I need to create other tags or is it that I put the `aws:cloudformation:stack-name` tag for every resource that accepts it? And I'm not sure where I should put the Name (optional) part?

For the folder name I've actually changed it in each command's branches but forgot to pull the changes to this branch. Thanks for flagging!

Reply:
The `FunctionName` has to be unique. You're absolutely right in your understanding/explanation of the relationship between the lambdas and the respective SNS. However, assume we have a `slack-bots-staging` and `slack-bots-production` stack. Each of those needs to have a different router lambda, which, like Dakota suggested will need to have a "stack-dependent" name to be unique, so the name will be based on whether the stack name is staging/production. You could simply have it be the `${name of the stack}-router-lambda`

As far as creating tags go, you wouldn't have to create them explicitly. AWS has some psuedo-parameters that can be referenced automatically in the CloudFormation: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html, so you can simply use a `cf.join` with the `cf.stackName` and a suffix, and that in itself will ensure the name is unique, even if multiple stacks are created.

Reply:
cdd3dafc1c3

I tried adding some tags but honestly not sure I did that right. I made a tag specifying what project it's part of, the permissions the Lambda has, and what triggers it.

Can I get some clarification on the tagging policy of HOT? I read up on how to do tagging (like this [AWS paper on tagging best practices](https://d1.awsstatic.com/whitepapers/aws-tagging-best-practices.pdf)) but it's mostly guidelines and seems like it's something different depending on the organisation.

I didn't create a `aws:cloudformation:stackName` tag since the [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-resource-tags.html) said it's automatically created and I didn't want to double up on tags but let me know if I need to anyway.